crates/wac-parser/src/resolution/types.rs

```diff
@@ -859,56 +914,73 @@ impl<'a> SubtypeChecker<'a> {
-                bail!("expected function parameter {i} to be named `{an}`, found name `{bn}`");
+                let (expected, found) = self.expected_found(an, bn);
+                bail!("expected function parameter {i} to be named `{expected}`, found name `{found}`");
             }
 
             self.value_type(*a, *b)
-                .with_context(|| format!("mismatched type for function parameter `{an}`"))?;
+                .with_context(|| format!("mismatched type for function parameter `{bn}`"))?;
         }
 
         match (&a.results, &b.results) {
-            (None, None) => {}
-            (None, Some(_)) => {
-                bail!("expected function without results, found function with results")
-            }
-            (Some(_), None) => {
-                bail!("expected function with results, found function without results")
-            }
+            (None, None) => return Ok(()),
             (Some(FuncResult::Scalar(a)), Some(FuncResult::Scalar(b))) => {
-                self.value_type(*a, *b)
-                    .context("mismatched type for function result")?;
-            }
-            (Some(FuncResult::Scalar(_)), Some(_)) => {
-                bail!("expected function with scalar result, found function with named results")
+                return self
+                    .value_type(*a, *b)
+                    .context("mismatched type for function result");
             }
             (Some(FuncResult::List(a)), Some(FuncResult::List(b))) => {
                 for (i, ((an, a), (bn, b))) in a.iter().zip(b.iter()).enumerate() {
                     if an != bn {
-                        bail!("expected function result {i} to be named `{an}`, found name `{bn}`");
+                        let (expected, found) = self.expected_found(an, bn);
+                        bail!("expected function result {i} to be named `{expected}`, found name `{found}`");
                     }
 
                     self.value_type(*a, *b)
-                        .with_context(|| format!("mismatched type for function result `{an}`"))?;
+                        .with_context(|| format!("mismatched type for function result `{bn}`"))?;
                 }
+
+                return Ok(());
             }
-            (Some(FuncResult::List(_)), Some(_)) => {
-                bail!("expected function with named results, found function with scalar result")
+            (Some(FuncResult::List(_)), Some(FuncResult::Scalar(_)))
+            | (Some(FuncResult::Scalar(_)), Some(FuncResult::List(_)))
+            | (Some(_), None)
+            | (None, Some(_)) => {
+                // Handle the mismatch below
             }
         }
 
-        Ok(())
+        let (expected, found) = self.expected_found(a, b);
+        match (&expected.results, &found.results) {
+            (Some(FuncResult::List(_)), Some(FuncResult::Scalar(_))) => {
+                bail!("expected function with named results, found function with scalar result")
```

Comment:
Not sure how I feel about these error messages. 

The usage of "scalar" might be confusing to some users. Am I understanding correctly that what is referred to here as "scalar" results are referred to in other codebases as "anonymous" types? 

I wonder if this an example of where being too specific makes things harder to understand. Perhaps we just need to label it as a type mismatch and move on.

Reply:
> The usage of "scalar" might be confusing to some users. Am I understanding correctly that what is referred to here as "scalar" results are referred to in other codebases as "anonymous" types?

I agree that "scalar" could be confusing here and perhaps simply saying "a single result type" would be better. I can update the messages.

It's not quite the same thing as an "anonymous" type, as that might convey meaning such as having an unnamed struct/record type in other languages; the type itself isn't anonymous, just the return value has no name associated with it, which is by far the most common case in WIT.

That's in contrast to named results in WIT, which (ironically) function more like an anonymous type as they are semantically equivalent to returning an anonymous record type.

> I wonder if this an example of where being too specific makes things harder to understand. Perhaps we just need to label it as a type mismatch and move on.

That's a totally reasonable position, but historically our error messages relating to these type checks have not been detailed, so I've erred on the side of providing as much information as possible.

Still, the action of the user would likely be to look at the signatures of the mismatched function and compare them visually, so perhaps the extra detail here wouldn't be necessary.

That said, I'm inclined to leave it as-is and see if anyone ever runs into these error messages in real-world scenarios.